trl/trainer/ppo_trainer.py

```diff
@@ -0,0 +1,579 @@
+# Copyright 2022 The HuggingFace Team. All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+from typing import Any, List, Optional, Union
+from accelerate import Accelerator
+
+from torch.optim import Adam
+import torch.nn as nn
+import torch
+import time
+import random
+import wandb
+
+from transformers import DataCollatorForLanguageModeling, PreTrainedTokenizer, PreTrainedTokenizerFast
+
+from trl.core import (logprobs_from_logits,
+                      whiten,
+                      clip_by_value,
+                      entropy_from_logits,
+                      flatten_dict,
+                      stats_to_np,
+                      stack_dicts,
+                      WANDB_PADDING)
+from trl.trainer import BaseTrainer, AdaptiveKLController, FixedKLController, LengthSampler
+from trl.models import PreTrainedModelWrapper, SUPPORTED_ARCHITECTURES
+
+
+class PPOTrainer(BaseTrainer):
+    """
+    The PPO_trainer uses Proximal Policy Optimization to optimise language models.
+    """
+
+    default_params = {
+        "lr": 1.41e-5,
+        "adap_kl_ctrl": True,
+        "init_kl_coef":0.2,
+        "target": 6,
+        "horizon":10000,
+        "gamma":1,
+        "lam":0.95,
+        "cliprange": .2,
+        "cliprange_value":.2,
+        "vf_coef":.1,
+        "batch_size": 256,
+        "forward_batch_size": 16,
+        "ppo_epochs": 4,
+        "log_with_wandb": True,
+    }
+
+    def __init__(
+        self,
+        model: PreTrainedModelWrapper, 
+        ref_model: PreTrainedModelWrapper, 
+        tokenizer: Union[PreTrainedTokenizer, PreTrainedTokenizerFast], 
+        dataloader: torch.utils.data.DataLoader,
+        **config
+    ):
+        """
+        Initialize PPOTrainer.
+
+        Args:
+            model (`PreTrainedModelWrapper`): 
+                Hugging Face transformer model with a value head.
+            ref_model (`PreTrainedModelWrapper`):
+                Hugging Face transformer GPT2 refrence model used for KL penalty
```

Comment:
nit: presumably it's not just for GPT-2 models right?
```suggestion
                Hugging Face transformer model with a casual language modelling head. Used for KL penalty
```